text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,607 @@
+performed.
+
+Should any check during compilation not pass a compiler error should be issued.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+<!-- Why should we *not* do this? -->
+As this feature should be mostly independent of other features the main drawback lies in the implementation and
+maintenance effort. This feature adds a new keyword which will need to be implemented not only in Rust but also in other
+tooling. The primary effort, however, lies in supporting this feature in the backends:
+- LLVM supports a `musttail` marker to indicate that TCO should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature, seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 30.03.23).
+- GCC does not support an equivalent `musttail` marker.
+- WebAssembly accepted tail-calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.
+
+Additionally, this proposal is limited to exactly matching function signatures which will *not* allow general tail-calls, however, the work towards this initial version is likely to be useful for a more comprehensive version.
+
+There is also an unwanted interaction between TCO and debugging. As TCO by design elides stack frames this information is lost during debugging, that is the parent functions and their local variable values are incomplete. As TCO provides a semantic guarantee of constant stack usage it is also not generally possible to disable TCO for debugging builds as then the stack could overflow. (Still maybe a compiler flag could be provided to temporarily disable TCO for debugging builds. As suggested [here](https://github.com/rust-lang/rfcs/pull/3407/files#r1159817279), another option would be special support for `become` by a debugger. With this support the debugger would keep track of the N most recent calls providing at least some context to the bug.)
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why is this design the best in the space of possible designs?
+This design is the best tradeoff between implementation effort and functionality, while also offering a good starting
+point toward further exploration of a more general implementation. To expand on this, compared to other options
+creating a function local scope with the use of `become` greatly reduces implementation effort. Additionally, limiting
+tail-callable functions to those with exactly matching function signatures enforces a common stack layout across all
+functions. This should in theory, depending on the backend, allow tail calls to be performed without any stack
+shuffling, indeed it might even be possible to do so for indirect calls or external functions.
```

Comment:
There's nothing in the RFC that prevents tail calls on indirect or external functions. These should be explicitly mentioned as supported.

Reply:
Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/377268eb567e9b4a435dcd187e7f83b7821f8fdc